File: Classes/BEMPathManager.m
//
//  BEMPathManager.m
//  CheckBox
//
//  Created by Bobo on 9/19/15.
//  Copyright (c) 2015 Boris Emorine. All rights reserved.
//

#import "BEMPathManager.h"

@implementation BEMPathManager

#pragma mark Paths

- (UIBezierPath *)pathForBox {
    UIBezierPath* path;
    switch (self.boxType) {
        case BEMBoxTypeSquare:
            path = [UIBezierPath bezierPathWithRoundedRect:CGRectMake(self.lineWidth / (CGFloat)2.0f, self.lineWidth / (CGFloat)2.0f, self.size - self.lineWidth, self.size - self.lineWidth) cornerRadius:self.cornerRadius];
            [path applyTransform:CGAffineTransformRotate(CGAffineTransformIdentity, (CGFloat)(M_PI * 2.5))];
            [path applyTransform:CGAffineTransformMakeTranslation(self.size, 0)];
            break;
            
        default: {
            CGFloat radius = (self.size / (CGFloat)2.0f) - (self.lineWidth / (CGFloat)2.0f);
            path = [UIBezierPath bezierPathWithArcCenter:CGPointMake(self.size / 2, self.size / 2)
                                                  radius: radius
                                              startAngle: - (CGFloat)(M_PI / 4.0)
                                                endAngle: (CGFloat)(2.0 * M_PI - M_PI / 4.0)
                                               clockwise:YES];
        }
            break;
    }
    return path;
}

- (UIBezierPath *)pathForCheckMark {
    UIBezierPath* checkMarkPath = [UIBezierPath bezierPath];
    
    [checkMarkPath moveToPoint: CGPointMake(self.size/(CGFloat)3.1578f, self.size/(CGFloat)2.0f)];
    [checkMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)2.0618f, self.size/(CGFloat)1.57894f)];
    [checkMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)1.3953f, self.size/(CGFloat)2.7272f)];
    
    if (self.boxType == BEMBoxTypeSquare) {
        // If we use a square box, the check mark should be a little bit bigger
        [checkMarkPath applyTransform:CGAffineTransformMakeScale(1.5, 1.5)];
        [checkMarkPath applyTransform:CGAffineTransformMakeTranslation(-self.size/4, -self.size/4)];
    }
    
    return checkMarkPath;
}

- (UIBezierPath *)pathForLongCheckMark {
    UIBezierPath* checkMarkPath = [UIBezierPath bezierPath];
    
    [checkMarkPath moveToPoint: CGPointMake(self.size/(CGFloat)3.1578f, self.size/(CGFloat)2.0f)];
    [checkMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)2.0618f, self.size/(CGFloat)1.57894f)];
    
    if (self.boxType == BEMBoxTypeSquare) {
        // If we use a square box, the check mark should be a little bit bigger
        [checkMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)1.2053f, self.size/(CGFloat)4.5272f)];
        [checkMarkPath applyTransform:CGAffineTransformMakeScale(1.5, 1.5)];
        [checkMarkPath applyTransform:CGAffineTransformMakeTranslation(-self.size/4, -self.size/4)];
    } else {
        [checkMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)1.1553f, self.size/(CGFloat)5.9272f)];
    }
    
    return checkMarkPath;
}

- (UIBezierPath *)pathForFlatCheckMark {
    UIBezierPath* flatCheckMarkPath = [UIBezierPath bezierPath];
    [flatCheckMarkPath moveToPoint: CGPointMake(self.size/4, self.size/2)];
    [flatCheckMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)2.0f, self.size/(CGFloat)2.0f)];
    [flatCheckMarkPath addLineToPoint: CGPointMake(self.size/(CGFloat)1.2f, self.size/(CGFloat)2.0f)];
    
    return flatCheckMarkPath;
}

@end

File: Classes/BEMPathManager.swift
//
//  BEMPathManager.swift
//  CheckBox
//
//  Created by Bobo on 9/19/15.
//  Copyright (c) 2015 Boris Emorine. All rights reserved.
//

import UIKit

final class BEMPathManager: NSObject {
    /** The paths are assumed to be created in squares.
     * This is the size of width, or height, of the paths that will be created.
     */
    var size: CGFloat = 0.0
    
    /** The width of the lines on the created paths. */
    var lineWidth: CGFloat = 0.0
    
    /** The corner radius of the path when the boxType is BEMBoxTypeSquare. */
    var cornerRadius: CGFloat = 0.0
    
    /** The type of box.
     * Depending on the box type, paths may be created differently
     * @see BEMBoxType
     */
    var boxType: BEMCheckBox.BoxType = .circle
    
    /** Returns a UIBezierPath object for the box of the checkbox
     * @returns The path of the box.
     */
    func pathForBox() -> UIBezierPath {
        var path: UIBezierPath
        
        switch boxType {
        case .square:
            let offset = lineWidth / 2.0
            let size = self.size - lineWidth
            
            path = UIBezierPath(
                roundedRect: CGRect(
                    x: 0.0, y: 0.0,
                    width: size, height: size),
                cornerRadius: cornerRadius)
            path.apply(CGAffineTransform.identity.rotated(by: CGFloat.pi * 2.5))
            path.apply(CGAffineTransform(translationX: size + offset, y: offset))
            
        default:
            let radius: CGFloat = (size - lineWidth) / 2
            
            path = UIBezierPath(
                arcCenter: CGPoint(x: size / 2, y: size / 2),
                radius: radius,
                startAngle: -.pi / 4,
                endAngle: 2 * .pi - .pi / 4,
                clockwise: true)
        }
        
        return path
    }
    
    /** Returns a UIBezierPath object for the checkmark of the checkbox
     * @returns The path of the checkmark.
     */
    func pathForCheckMark() -> UIBezierPath {
        let checkMarkPath = UIBezierPath()
        
        checkMarkPath.move(to: CGPoint(x: size / 3.1578, y: size / 2))
        checkMarkPath.addLine(to: CGPoint(x: size / 2.0618, y: size / 1.57894))
        checkMarkPath.addLine(to: CGPoint(x: size / 1.3953, y: size / 2.7272))
        
        if boxType == .square {
            // If we use a square box, the check mark should be a little bit bigger
            checkMarkPath.apply(CGAffineTransform(scaleX: 1.5, y: 1.5))
            checkMarkPath.apply(CGAffineTransform(translationX: -size / 4, y: -size / 4))
        }
        
        return checkMarkPath
    }
    
    /** Returns a UIBezierPath object for an extra long checkmark which is in contact with the box.
     * @returns The path of the checkmark.
     */
    func pathForLongCheckMark() -> UIBezierPath {
        let checkMarkPath = UIBezierPath()
        
        checkMarkPath.move(to: CGPoint(x: size / 3.1578, y: size / 2))
        checkMarkPath.addLine(to: CGPoint(x: size / 2.0618, y: size / 1.57894))
        
        if boxType == .square {
            // If we use a square box, the check mark should be a little bit bigger
            checkMarkPath.addLine(to: CGPoint(x: size / 1.2053, y: size / 4.5272))
            checkMarkPath.apply(CGAffineTransform(scaleX: 1.5, y: 1.5))
            checkMarkPath.apply(CGAffineTransform(translationX: -size / 4, y: -size / 4))
        } else {
            checkMarkPath.addLine(to: CGPoint(x: size / 1.1553, y: size / 5.9272))
        }
        
        return checkMarkPath
    }
    
    /** Returns a UIBezierPath object for the flat checkmark of the checkbox
     * @see BEMAnimationTypeFlat
     * @returns The path of the flat checkmark.
     */
    func pathForFlatCheckMark() -> UIBezierPath {
        let flatCheckMarkPath = UIBezierPath()
        flatCheckMarkPath.move(to: CGPoint(x: size / 4, y: size / 2))
        flatCheckMarkPath.addLine(to: CGPoint(x: size / 2, y: size / 2))
        flatCheckMarkPath.addLine(to: CGPoint(x: size / 1.2, y: size / 2))
        
        return flatCheckMarkPath
    }
}
